### ISSUE TITLE
strange error about invariance

### ISSUE THREAD

**Issue by dearoneesama** [microsoft/pyright#2081]
**Describe the bug**
Pyright reports new error after updating to the latest version. I have no clue what it is about exactly. Looks like something about invariance.

**To Reproduce**
Type code

**Expected behavior**
No error as in the previous version

**Screenshots or Code**
```python
T = TypeVar('T')
Tco = TypeVar('Tco', covariant=True)
Tcontra = TypeVar('Tcontra', contravariant=True)

class BrokenClass(Generic[T]):
    def broken(self):
        p0 = cast(IrrelevantClass[T], object())
        p1 = cast(Callable[[T], object], object())
        func(p0, p1)

def func(
    p0: IrrelevantClass[T],
    p1: Callable[[T], object],
): ...

class IrrelevantClass(Generic[T]): ...
```

error:
```
error: Argument of type "IrrelevantClass[T@BrokenClass]" cannot be assigned to parameter "p0" of type "IrrelevantClass[T@func]" in function "func"
    TypeVar "T@IrrelevantClass" is invariant
      Type "T@BrokenClass" cannot be assigned to type "T@BrokenClass" (reportGeneralTypeIssues)
```

* If I changed `class IrrelevantClass(Generic[T])` to `Tco` or `Tcontra`, there was no error
* If I changed the type of `p1` to something else that is not callable of `T`, there was no error

**VS Code extension or command-line**
1.1.156

**Comment by erictraut** [comment]
Thanks for the bug report, and apologies for the regression. I've fixed the bug and added a new unit test based on your code sample so we don't have a similar regression in the future.

**Comment by erictraut** [comment]
This fix is included in pyright 1.1.57, which I just published. It will also be included in the next release of pylance.
